closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -380,30 +401,91 @@ JNIEXPORT jboolean JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_isOpenSSLFIP
     return OSSL_IS_FIPS;
 }
 
-/*
- * Class:     jdk_crypto_jniprovider_NativeCrypto
- * Method:    loadCrypto
- * Signature: (Z)J
- */
-JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
-  (JNIEnv *env, jclass thisObj, jboolean trace)
-{
-    typedef const char* OSSL_version_t(int);
+/* Unload the crypto library */
+static void unload_crypto_library(void *crypto_library) {
+    #if defined (_WIN32)
+    FreeLibrary(crypto_library);
+    #else
+    (void)dlclose(crypto_library);
+    #endif
+}
 
-    /* Determine the version of OpenSSL. */
-    OSSL_version_t* OSSL_version;
-    const char * openssl_version;
-    jlong ossl_ver = 0;
+/* Find the symbol in the crypto library (return NULL if not found) */
+static void * find_crypto_symbol(void *crypto_library, const char *symname) {
+    #if defined (_WIN32)
+    return GetProcAddress(crypto_library, symname);
+    #else
+    return dlsym(crypto_library, symname);
+    #endif
+}
 
-    /* Load OpenSSL Crypto library */
-    crypto_library = load_crypto_library(trace);
-    if (NULL == crypto_library) {
-        if (trace) {
-            fprintf(stderr, "Error loading OpenSSL: FAILED TO LOAD OPENSSL CRYPTO LIBRARY\n");
-            fflush(stderr);
+void log_crypto_library_path(jboolean traceEnabled, void *crypto_library, char *message) {
+    if (traceEnabled && (NULL != crypto_library)) {
+
+    #if defined(_AIX)
+        // Initialize the buffer with maximum size for L_GETINFO.
+        char *buffer = (char *) malloc(DLFCN_LDINFO_SIZE);
+        int rc = -1;
+        // Get the list of all object files loaded by this process
+        rc = loadquery(L_GETINFO, buffer, DLFCN_LDINFO_SIZE);
+        if (NULL == buffer) {
+            return;
         }
-        return -1;
+
+        // Parse the list of all object files and print the OPENSSL library path.
+        if (rc == 0) {
+            char *buf = buffer;
+            struct ld_info* cur_info = NULL;
+            do {
+                cur_info = (struct ld_info*) buf;
+                char *path = cur_info->ldinfo_filename;
+                char *member_name = path + strlen(cur_info->ldinfo_filename) + 1;
+                if ((*member_name != '\0') && (strstr(path, "libcrypto") != NULL)) {
+                    fprintf(stdout, "%s: %s(%s)\n", message, path, member_name);
+                    fflush(stdout);
+                    free(buffer);
+                    return;
+                }
+                buf += cur_info->ldinfo_next;
+            } while (cur_info->ldinfo_next != 0);
+        }
+        free(buffer);
+
+    #elif defined(__APPLE__)
+        // Since we know the image we want will always be near the end of the list, start there and go backwards.
+        for (uint32_t i = (_dyld_image_count() - 1); i >= 0; i--) {
+            const char *image_name = _dyld_get_image_name(i);
+
+            // Why dlopen doesn't effect _dyld stuff: if an image is already loaded, it returns the existing handle.
```

Comment:
This needs to be a C style comment; I think you meant "affect" instead of "effect".

Reply:
Fixed.